velox/common/memory/AllocationPool.cpp

```diff
@@ -29,31 +29,49 @@ void AllocationPool::clear() {
   largeAllocations_.clear();
 }
 
-char* AllocationPool::allocateFixed(uint64_t bytes) {
-  VELOX_CHECK_GT(bytes, 0, "Cannot allocate zero bytes");
+namespace {
 
-  auto numPages = bits::roundUp(bytes, memory::AllocationTraits::kPageSize) /
+uint64_t getNumPages(uint64_t bytes) {
+  return bits::roundUp(bytes, memory::AllocationTraits::kPageSize) /
       memory::AllocationTraits::kPageSize;
+}
+
+} // namespace
+
+char* AllocationPool::allocateFixed(uint64_t bytes, int32_t alignment) {
+  VELOX_CHECK_GT(bytes, 0, "Cannot allocate zero bytes");
+  VELOX_CHECK_EQ(
+      __builtin_popcount(alignment), 1, "Alignment can only be power of 2");
 
   // Use contiguous allocations from mapped memory if allocation size is large
-  if (numPages > pool_->largestSizeClass()) {
+  if (auto numPages = getNumPages(bytes + alignment - 1);
+      numPages > pool_->largestSizeClass()) {
     auto largeAlloc = std::make_unique<memory::ContiguousAllocation>();
     pool_->allocateContiguous(numPages, *largeAlloc);
     largeAllocations_.emplace_back(std::move(largeAlloc));
     auto res = largeAllocations_.back()->data<char>();
     VELOX_CHECK_NOT_NULL(
         res, "Unexpected nullptr for large contiguous allocation");
+    // Should be at page boundary and always aligned.
+    VELOX_CHECK_EQ(reinterpret_cast<uintptr_t>(res) % alignment, 0);
     return res;
   }
 
-  if (availableInRun() < bytes) {
-    newRunImpl(numPages);
+  if (availableInRun() == 0) {
+    newRunImpl(getNumPages(bytes + alignment - 1));
```

Comment:
Should we compute numPages at the beginning of this method, store it in a temp variable and reuse?